hw/src/trezor.rs

```diff
@@ -0,0 +1,431 @@
+						// so v' = v + 2 * chain_id + 35, but code further down the
+						// pipeline will already do this transformation, so remove it here
+						Ok(Signature::from_rsv(&r, &s, (v - (35 + 2 * c_id as u32)) as u8))
+					} else {
+						// If there isn't a chain_id, v will be returned as v + 27
+						Ok(Signature::from_rsv(&r, &s, (v - 27) as u8))
+					}
+				}
+			}
+			MessageType::MessageType_Failure => Err(Error::Protocol("Last message sent to Trezor failed")),
+			_ => Err(Error::Protocol("Unexpected response from Trezor device.")),
+		}
+	}
+
+	fn send_device_message(&self, device: &hidapi::HidDevice, msg_type: &MessageType, msg: &Message) -> Result<usize, Error> {
+		let msg_id = *msg_type as u16;
+		let mut message = msg.write_to_bytes()?;
+		let msg_size = message.len();
+		let mut data = Vec::new();
+		let hid_version = self.probe_hid_version(device)?;
+		// Magic constants
+		data.push('#' as u8);
+		data.push('#' as u8);
+		// Convert msg_id to BE and split into bytes
+		data.push(((msg_id >> 8) & 0xFF) as u8);
+		data.push((msg_id & 0xFF) as u8);
+		// Convert msg_size to BE and split into bytes
+		data.push(((msg_size >> 24) & 0xFF) as u8);
+		data.push(((msg_size >> 16) & 0xFF) as u8);
+		data.push(((msg_size >> 8) & 0xFF) as u8);
+		data.push((msg_size & 0xFF) as u8);
+		data.append(&mut message);
+		while data.len() % 63 > 0 {
+			data.push(0);
+		}
+		let mut total_written = 0;
+		for chunk in data.chunks(63) {
+			let mut padded_chunk = match hid_version {
+				HidVersion::V1 => vec!['?' as u8],
+				HidVersion::V2 => vec![0, '?' as u8],
+			};
+			padded_chunk.extend_from_slice(&chunk);
+			total_written += device.write(&padded_chunk)?;
+		}
+		Ok(total_written)
+	}
+
+	fn probe_hid_version(&self, device: &hidapi::HidDevice) -> Result<HidVersion, Error> {
+		let mut buf2 = [0xFFu8; 65];
+		buf2[0] = 0;
+		buf2[1] = 63;
+		let mut buf1 = [0xFFu8; 64];
+		buf1[0] = 63;
+		if device.write(&buf2)? == 65 {
+			Ok(HidVersion::V2)
+		} else if device.write(&buf1)? == 64 {
+			Ok(HidVersion::V1)
+		} else {
+			Err(Error::Usb("Unable to determine HID Version"))
+		}
+	}
+
+	fn read_device_response(&self, device: &hidapi::HidDevice) -> Result<(MessageType, Vec<u8>), Error> {
+		let protocol_err = Error::Protocol(&"Unexpected wire response from Trezor Device");
+		let mut buf = vec![0; 64];
+
+		let first_chunk = device.read_timeout(&mut buf, 300_000)?;
+		if first_chunk < 9 || buf[0] != '?' as u8 || buf[1] != '#' as u8 || buf[2] != '#' as u8 {
+			return Err(protocol_err);
+		}
+		let msg_type = MessageType::from_i32(((buf[3] as i32 & 0xFF) << 8) + (buf[4] as i32 & 0xFF)).ok_or(protocol_err)?;
+		let msg_size = ((buf[5] as u32 & 0xFF) << 24) + ((buf[6] as u32 & 0xFF) << 16) + ((buf[7] as u32 & 0xFF) << 8) + (buf[8] as u32 & 0xFF);
+		let mut data = Vec::new();
+		data.extend_from_slice(&buf[9..]);
+		while data.len() < (msg_size as usize) {
+			device.read_timeout(&mut buf, 10_000)?;
+			data.extend_from_slice(&buf[1..]);
```

Comment:
How do we know that the whole buffer is filled?

Reply:
Such is the protocol of the device, if it were to return fewer bytes, that would be a protocol error. The bytes returned from this function would be somewhat "mangled" and thus unable to parse into a protobuf message later, resulting in a protocol error being thrown.

Reply:
why does it always discard the first read byte on each iteration of the loop?

Reply:
The first byte of each 64 byte increment from the device is always `32` as an error check to make sure each send isn't messed up. Ideally, each time new data comes it, it's worthwhile to authenticate.

Reply:
Yeah, part of the protocol that the first byte is '?', maybe I should throw in a quick check that it is that? If the data is wrong and it can't be parsed into whatever protobuf it's supposed to be, it will throw an error anyway though? Not exactly sure what adding a check gives us.